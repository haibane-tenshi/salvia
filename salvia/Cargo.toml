[package]
name = "salvia"
version = "0.1.0"
edition = "2018"

description = "Incremental computing brought to async Rust"
authors = ["haibane_tenshi <obscure.sanctum@gmail.com>"]
license = "MIT OR Apache-2.0"
readme = "README.md"
keywords = ["incremental"]
categories = ["asynchronous", "caching"]

[features]
nightly = []

[package.metadata.docs.rs]
features = ["async-trait", "nightly"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
salvia_macro = { path = '../salvia_macro' }

anymap2 = "0.13.0"
futures = "0.3.13"
tokio = { version = "1.4.0", features = ["rt", "sync", "macros"] }
parking_lot = "0.11.1"
bitvec = "0.22.3"
async-trait = { version = "0.1.50", optional = true }
tracing = { version = "0.1.29", optional = true }

[dev-dependencies]
# Need this available for doc-tests.
# Macro expansion relies on `salvia` name to be available in global scope, so we need to alias ourselves here.
# Not sure if there is a better way.
salvia = { path = '.' }

tokio = { version = "1.4.0", features = ["rt-multi-thread", "sync", "macros"] }
async-trait = "0.1.50"
async-recursion = "0.3.2"
nom = "7.1.0"
