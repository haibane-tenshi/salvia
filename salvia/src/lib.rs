//! Incremental computing brought to async Rust.
//!
//! **Obligatory warning**
//!
//! This crate is highly experimental.
//! Unless you absolutely need `async` consider using [`salsa`][salsa] instead.
//!
//! # Contents
//!
//! * [Primer](#primer)
//! * [Quick start](#quick-start)
//! * [Required trait bounds](#required-trait-bounds)
//! * [How it works](#how-it-works)
//! * [Features](#features)
//! * [Known limitations](#known-limitations)
//! * [Similar projects](#similar-projects)
//!
//! # Primer
//!
//! * [Async in Rust](https://rust-lang.github.io/async-book/)
//! * [Incremental computing](https://en.wikipedia.org/wiki/Incremental_computing)
//! * [Adapton][adapton] - research initiative for incremental computing
//! * [`salsa`][salsa] - incremental computation framework for sync Rust
//!
//! `salvia` allows you to define *queries* (functions which values will be cached) and *inputs*
//! ("functions" which value is set directly by user).
//! Upon execution, queries record which other queries or inputs it called and can avoid
//! recalculation when none of those values change.
//!
//! `salvia` was inspired by `salsa` library.
//! As a major difference from it, this crate was designed to work in async Rust from the start
//! and had to make some concessions to achieve that.
//!
//! Currently, only [`tokio`][tokio] runtime is supported.
//!
//! [tokio]: https://github.com/tokio-rs/tokio
//! [salsa]: https://github.com/salsa-rs/salsa
//! [adapton]: http://adapton.org/
//!
//! # Quick start
//!
//! 1.  *Define inputs.*
//!
//!     Input acts as *handle to storage* for values of specific type.
//!     Inputs can be explicitly set, so they serve as a primary entry point for user data.
//!
//!     Conceptually inputs are implemented as traits, so we will need a *host* type
//!     to attach input to.
//!
//!     ```
//!     #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, salvia::InputAccess)]
//!     //                                                        ^^^^^^^^^^^
//!     // derive `InputAccess` to implement `get` and `set` functions on type
//!     enum Team {
//!         Blue,
//!         Red,
//!     }
//!
//!     #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
//!     struct Score(u32);
//!
//!     // implement `Input` trait to associate type with input
//!     impl salvia::Input<Score> for Team {
//!     //                 ^^^^^
//!     // generic param indicates the type of stored value
//!
//!         fn initial(&self) -> Score {
//!         // ^^^^^^^
//!         // `initial` function will be called to initialize the storage if needed
//!
//!             Score(0)
//!         }
//!     }
//!     ```
//!
//!     You can implement `Input` multiple times using different `Value` types.
//!
//!     Deriving [`InputAccess`](macro@InputAccess) implements two (generic) functions
//!     `get` and `set` on `Host`.
//!     Unlike normal derive macros `InputAccess` does not implement a trait,
//!     it rather implements functions **directly on the type**.
//!     For reasons behind this design (and alternatives) see [`input`][node::input] module.
//!
//! 2.  *Define queries.*
//!
//!     Queries are *pure functions* which can call other queries or inputs.
//!     Do not forget to make them async!
//!
//!     ```
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, salvia::InputAccess)]
//!     # enum Team {
//!     #     Blue,
//!     #     Red,
//!     # }
//!     #
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
//!     # struct Score(u32);
//!     #
//!     # impl salvia::Input<Score> for Team {
//!     #     fn initial(&self) -> Score {
//!     #         Score(0)
//!     #     }
//!     # }
//!     #
//!     #[salvia::query]
//!     async fn leader(cx: &salvia::QueryContext) -> Option<Team> {
//!         //          ^^
//!         // query should take `&QueryContext` as last parameter...
//!         use std::cmp::Ordering;
//!
//!         let red: Score = Team::Red.get(cx).await;
//!         //                             ^^
//!         // ... and just pass it into any other queries it calls
//!
//!         let blue: Score = Team::Blue.get(cx).await;
//!         //                           ^^^
//!         // `get` is a query which allows you to access value of an input
//!         // it has signature of `async fn get<T>(self, &QueryContext) -> T`
//!         // you can use it similarly to `Iterator::collect()` by specifying expected type
//!
//!         match red.cmp(&blue) {
//!             Ordering::Less => Some(Team::Blue),
//!             Ordering::Equal => None,
//!             Ordering::Greater => Some(Team::Red),
//!         }
//!     }
//!     ```
//!
//!     The macro is not picky and supports transformation on all kinds of functions.
//!     Check [`#[query]`](`query`) attribute documentation to learn more
//!     about its usage.
//!
//! 3.  *Create runtime*
//!
//!     Runtime holds references to internal storage and mediates synchronization.
//!
//!     You can create one by calling to [`Runtime::new`] from inside a task:
//!
//!     ```
//!     #[tokio::main]
//!     async fn main() {
//!         let rt = salvia::Runtime::new().await;
//!     }
//!     ```
//!
//!     `new` is guaranteed to return `Poll::Ready`, `async` is just a reminder that you have
//!     to call from within tokio context,
//!     runtime needs access to it to set up internal tasks.
//!
//!     Alternatively, in synchronous context you can use [`Runtime::with_tokio`]:
//!
//!     ```
//!     # fn main() -> Result<(), Box<dyn std::error::Error>> {
//!     let tokio = tokio::runtime::Runtime::new()?;
//!     let rt = salvia::Runtime::with_tokio(&tokio);
//!     # Ok(())
//!     # }
//!     ```
//!
//! 4. *Run*
//!
//!     Now we can update inputs
//!
//!     ```
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, salvia::InputAccess)]
//!     # enum Team {
//!     #     Blue,
//!     #     Red,
//!     # }
//!     #
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
//!     # struct Score(u32);
//!     #
//!     # impl salvia::Input<Score> for Team {
//!     #     fn initial(&self) -> Score {
//!     #         Score(0)
//!     #     }
//!     # }
//!     #
//!     # #[salvia::query]
//!     # async fn leader(cx: &salvia::QueryContext) -> Option<Team> {
//!     #     use std::cmp::Ordering;
//!     #
//!     #     let red: Score = Team::Red.get(cx).await;
//!     #     let blue: Score = Team::Blue.get(cx).await;
//!     #
//!     #     match red.cmp(&blue) {
//!     #         Ordering::Less => Some(Team::Blue),
//!     #         Ordering::Equal => None,
//!     #         Ordering::Greater => Some(Team::Red),
//!     #     }
//!     # }
//!     #
//!     # #[tokio::main]
//!     # async fn main() {
//!     #     let rt = salvia::Runtime::new().await;
//!     #
//!     rt.mutate(|cx| async move {
//!         //     ^^
//!         // `Runtime::mutate` gives you access to `InputContext`
//!
//!         Team::Red.set(Score(1), &cx).await;
//!         //        ^^^
//!         // `set` allows you to mutate an input
//!         // it has signature of `async fn set<T>(self, value: T, cx: &InputContext)`
//!         //                                                           ^^^^^^^^^^^^
//!         // note that setters take `&InputContext`, a different context type from queries
//!     }).await;
//!     # }
//!     ```
//!
//!     or run queries
//!
//!     ```
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, salvia::InputAccess)]
//!     # enum Team {
//!     #     Blue,
//!     #     Red,
//!     # }
//!     #
//!     # #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
//!     # struct Score(u32);
//!     #
//!     # impl salvia::Input<Score> for Team {
//!     #     fn initial(&self) -> Score {
//!     #         Score(0)
//!     #     }
//!     # }
//!     #
//!     # #[salvia::query]
//!     # async fn leader(cx: &salvia::QueryContext) -> Option<Team> {
//!     #     use std::cmp::Ordering;
//!     #
//!     #     let red: Score = Team::Red.get(cx).await;
//!     #     let blue: Score = Team::Blue.get(cx).await;
//!     #
//!     #     match red.cmp(&blue) {
//!     #         Ordering::Less => Some(Team::Blue),
//!     #         Ordering::Equal => None,
//!     #         Ordering::Greater => Some(Team::Red),
//!     #     }
//!     # }
//!     #
//!     # #[tokio::main]
//!     # async fn main() {
//!     #     let rt = salvia::Runtime::new().await;
//!     #
//!     #     rt.mutate(|cx| async move {
//!     #         Team::Red.set(Score(1), &cx).await;
//!     #     }).await;
//!     let leader = rt.query(|cx| async move {
//!         //                 ^^
//!         // `Runtime::query` gives you access to `QueryContext`
//!
//!         leader(&cx).await
//!         //     ^
//!         // both `query` and `mutate` provide context by value,
//!         // so you need to reference it
//!     }).await;
//!
//!     assert_eq!(leader, Some(Team::Red));
//!     # }
//!     ```
//!
//!     Note that both [`Runtime::query`] and [`Runtime::mutate`] provide appropriate context
//!     *by value*, which is inconsistent with normal queries.
//!     Unfortunately this is caused by a long-standing issue between closures, async and lifetimes:
//!     currently there is no way to pass a reference into async closure and properly constrain
//!     lifetimes so it can be consumed by other functions :(
//!
//! # Required trait bounds
//!
//! ## Core traits
//!
//! All involved value types are required to implement selection of traits:
//!
//! *   `Clone` - required to duplicate values to rerun queries and to clone cache contents
//! *   `Eq` - required to determine if queries are outdated and need to be rerun
//! *   `Send` - required to send between tasks (which potentially can send them between threads)
//! *   `Sync` - interestingly, `Sync` is not strictly required (in fact early prototypes didn't
//!     use it at all), however it helps to avoid excess cloning in multiple cases
//! *   `'static` - required to persist values and to send them between tasks
//!
//! `Clone + Eq + Send + Sync + 'static` can be quite a mouthful to type,
//! so `salvia` provides a special [`Stashable`] trait which can be used as a convenient shortcut.
//!
//! ## Additional traits
//!
//! Query parameters and input hosts are also required to implement `Hash`.
//!
//! `salvia`'s unit of operation is *bound query*.
//! *Bound query* is a combination of query with its parameter values.
//! To draw an analogy if normal query is like a function
//! then *bound query* is a closure which captures all parameters and forwards them to the function:
//!
//! ```
//! fn query(a: u32, b: f64, c: &str) {}
//!
//! let a = 2;
//! let b = 3.5;
//! let c = "beautiful tree";
//!
//! let bound_query = move || {
//!     // a, b, c is captured here
//!     query(a, b, c)
//! };
//! ```
//!
//! Why is this important?
//! *Bound queries* are handled independently from each other and have separate caches.
//! Which brings forth a question: how does framework finds its way back to a specific cache?
//! Queries are represented as different functions so we can statically distinguish between them,
//! but then we also need to do the same for parameter values.
//!
//! `salvia` opts for simple solution: `HashMap`.
//! The observation here is that types you can put into framework are "data-like",
//! so they almost never have a reason to not implement `Hash`.
//!
//! It is possible to avoid this extra bound at the cost of linear lookup,
//! but that hurts every other use case.
//!
//! # How it works
//!
//! On important implementation details and how it affects runtime characteristics see
//! [`runtime`][runtime] module-level documentation.
//!
//! # Features
//!
//! This crate provides following features:
//!
//! *   `async-trait` - enable [`InputAccess`](trait@InputAccess) trait,
//!     pulls in [`async-trait`][docs.rs:async_trait] crate.
//! *   `tracing` - enable internal logging via [`tracing`][docs.rs:tracing] crate.
//!     It gets quite verbose, intended for debugging of `salvia` itself.
//!     Log contents are not part of SemVer guarantees.
//!     See [CONTRIBUTING][repo:contributing] doc for more details on logging levels.
//!
//!     You **should not** use it in production or enable in your lib:
//!     it silently modifies definition of [`Stashable`] by adding [`Debug`] trait.
//! *   `nightly` - enable some unstable features.
//!     At the moment it just enables [`doc_cfg`][rustdoc:doc_cfg],
//!     so it is only useful when compiling documentation.
//!
//! [repo:contributing]: https://github.com/haibane-tenshi/salvia/blob/main/CONTRIBUTING.md#using-tracing
//! [docs.rs:async_trait]: https://docs.rs/async-trait/latest/async_trait/
//! [docs.rs:tracing]: https://docs.rs/tracing/latest/tracing/
//! [rustdoc:doc_cfg]: https://doc.rust-lang.org/rustdoc/unstable-features.html#doccfg-recording-what-platforms-or-features-are-required-for-code-to-be-present
//!
//! # Known limitations
//!
//! *   Only `tokio` runtime is supported.
//!
//!     But it should be possible to support other runtimes in the future as long as there are ways
//!     to abstract over them.
//!
//! *   No proper `Input` trait.
//!
//!     This is caused by lack of certain features in Rust type system, so it can be lifted in
//!     distant future.
//!     See [`input`][node::input] module documentation for details.
//!
//! *   No serialization/deserialization capabilities.
//!
//!     [`anymap2`] crate is used internally which makes it unclear to how even approach
//!     this topic.
//!
//!     See [DESIGN_DOC][repo:design_doc] for list of possible approaches.
//!
//! *   No node deallocation.
//!
//!     Currently nodes once started terminate only when runtime is dropped.
//!     That is a valid strategy, but can cause memory issues when too many queries with diverse
//!     sets of arguments are called.
//!
//!     It should possible to implement some kind of cache invalidation system, but within
//!     the current architecture this will require periodically waking nodes at masse.
//!     This is against the core idea of laziness, so it is not implemented at the moment.
//!
//! *   No *external* synchronization mechanisms.
//!
//!     `salvia` provides only two *internal* synchronization mechanisms:
//!
//!     *   `Runtime::query` guarantees that all queries executed within it agree on which input
//!         values they see.
//!     *   `Runtime::mutate` guarantees that passed closure is executed atomically w.r.t. to other
//!         calls to `mutate`.
//!
//!     However there is no synchronization means between individual calls
//!     to `query`/`mutate`.
//!
//!     You can sequence them by just polling one to completion before other:
//!
//!     ```
//!     # #[tokio::main]
//!     # async fn main() {
//!     # let rt = salvia::Runtime::new().await;
//!     rt.mutate(|cx| async move {
//!         // change inputs...
//!     }).await; // this await makes sure that mutation went through
//!
//!     rt.query(|cx| async move {
//!         // query values...
//!     }).await; // this await makes sure that query has finished
//!     # }
//!     ```
//!
//!     However this does not guarantee that query will see exact same values that you just set:
//!     indeed it is possible for some other task to change an input value after `mutate`
//!     finishes but before `query` starts!
//!     Similarly, if two tasks try to write to the same input concurrently there are no guarantees
//!     about if and which value will be observed by queries.
//!     You can be sure that one value is going to persist (the one that arrived last),
//!     but it can be either of the two.
//!
//!     If you are in need for more sophisticated dataflow control you should lean into normal
//!     async synchronization methods (manager tasks, semaphores, etc).
//!
//! [repo:design_doc]: https://github.com/haibane-tenshi/salvia/blob/main/DESIGN_DOC.md#deserialization
//!
//! # Similar projects
//!
//! * [`salsa`][docs.rs:salsa]
//!
//!     Similarities:
//!     * Both are based on pure functions
//!     * Both use query/input structure
//!
//!     Key differences:
//!     * `salsa` is sync, `salvia` is async
//!     *   `salsa`'s queries are resolved statically via traits,
//!         i.e. given a runtime you can only run queries implemented on it.
//!
//!         `salvia`'s queries are resolved dynamically, i.e. you can run any query in any runtime.
//!
//! * [`futures_signals`][docs.rs:futures_signals]
//!
//!     Similarities:
//!     * Both deal with async
//!
//!     Key differences:
//!     *   `futures_signals` is a "push-based" framework, i.e.
//!         * computations are eager (introducing change causes computations),
//!         * computations are effectfull (operate with side-effects).
//!
//!         `salvia` is "pull-based" framework, i.e.
//!         *   computations are lazy (introducing change nas no extra consequences;
//!             performed only on request),
//!         * computations are pure (don't produce side-effects).
//!
//!     *   `salvia` cares about result consistency: putting same inputs into computation
//!         necessarily produces the same outcome.
//!
//!
//! [docs.rs:salsa]: https://docs.rs/salsa/latest/salsa/
//! [docs.rs:futures_signals]: https://docs.rs/futures-signals/latest/futures_signals/
#![warn(missing_docs)]
#![allow(dead_code)]
#![cfg_attr(feature = "nightly", feature(doc_cfg))]

#[allow(unused_imports)]
use std::fmt::Debug;

pub use node::input::{Input, InputAccess};
pub use node::query::query;
pub use runtime::{InputContext, QueryContext, Runtime};

macro_rules! info {
    ($($t:tt)*) => {
        #[cfg(feature = "tracing")]
        ::tracing::info!($($t)*)
    }
}

macro_rules! debug {
    ($($t:tt)*) => {
        #[cfg(feature = "tracing")]
        ::tracing::debug!($($t)*)
    }
}

macro_rules! debug_span {
    ($($t:tt)*) => {{
        #[cfg(feature = "tracing")]
        ::tracing::debug_span!($($t)*)
    }}
}

pub mod node;
pub mod runtime;

pub(crate) mod time;
pub(crate) mod tracing;

#[doc(hidden)]
pub mod macro_support;

/// A shortcut for `Clone + Eq + Send + Sync + 'static`.
///
/// This trait is an umbrella for all trait bounds required for value types
/// flowing in or out of the framework.
/// See [crate level](self#core-traits) documentation for more details.
///
/// The purpose of this trait is to reduce boilerplate and chances of accidentally forgetting some
/// in generic context.
/// It is recommended to use the trait over specifying every bound individually.
#[cfg(not(feature = "tracing"))]
pub trait Stashable: Clone + Eq + Send + Sync + 'static {}

/// A shortcut for `Debug + Clone + Eq + Send + Sync + 'static`.
///
/// This trait is an umbrella for all trait bounds required for value types
/// flowing in or out of the framework.
/// See [crate level](self#core-traits) documentation for more details.
///
/// The purpose of this trait is to reduce boilerplate and chances of accidentally forgetting some
/// in generic context.
/// It is recommended to use the trait over specifying every bound individually.
///
/// **Warning**: this variant of trait is defined under `tracing` feature and is different
/// from normal definition (+ `Debug`).
#[cfg(feature = "tracing")]
pub trait Stashable: Debug + Clone + Eq + Send + Sync + 'static {}

#[cfg(not(feature = "tracing"))]
impl<T> Stashable for T where T: Clone + Eq + Send + Sync + 'static {}

#[cfg(feature = "tracing")]
impl<T> Stashable for T where T: Debug + Clone + Eq + Send + Sync + 'static {}

// Test examples from README.
#[doc = include_str!("../README.md")]
#[doc(hidden)]
const _: () = ();
