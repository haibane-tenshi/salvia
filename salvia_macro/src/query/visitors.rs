use proc_macro2::Span;
use syn::visit::Visit;
use syn::TypeImplTrait;

#[derive(Default)]
pub struct ImplTraitVisitor(pub Vec<Span>);

impl<'ast> Visit<'ast> for ImplTraitVisitor {
    fn visit_type_impl_trait(&mut self, node: &'ast TypeImplTrait) {
        use syn::spanned::Spanned;

        self.0.push(node.span())
    }
}
