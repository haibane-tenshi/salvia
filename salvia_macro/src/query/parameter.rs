use proc_macro2::Span;
use syn::{FnArg, Ident, PatIdent, Receiver, Type};

pub enum ReceiverIdent {
    Receiver(Receiver),
    PatIdent(PatIdent),
}

pub struct Parameters {
    pub receiver: Option<(ReceiverIdent, Type)>,
    pub common: Vec<(PatIdent, Type)>,
    pub context: (PatIdent, Type),
}

pub enum ParameterError {
    MissingContext,
    UnexpectedSelf(Span),
    Attribute(Span),
    PatternDestructuring(Span),
    WildcardPattern(Span),
    ImplTrait(Span),
}

impl Parameters {
    pub fn from_inputs(iter: impl IntoIterator<Item = FnArg>) -> Result<Self, ParameterError> {
        use syn::spanned::Spanned;
        use syn::{Pat, PatType};

        fn process_pat(pat: PatType) -> Result<(PatIdent, Type), ParameterError> {
            use super::visitors::ImplTraitVisitor;
            use syn::visit::Visit;

            if let Some(attr) = pat.attrs.first() {
                return Err(ParameterError::Attribute(attr.span()));
            }

            let ty = *pat.ty;
            let mut visitor = ImplTraitVisitor::default();
            visitor.visit_type(&ty);

            if let Some(span) = visitor.0.into_iter().next() {
                return Err(ParameterError::ImplTrait(span));
            }

            let ident = match *pat.pat {
                Pat::Ident(pat) => {
                    if let Some(attr) = pat.attrs.first() {
                        return Err(ParameterError::Attribute(attr.span()));
                    }

                    if let Some(subpat) = &pat.subpat {
                        return Err(ParameterError::PatternDestructuring(subpat.1.span()));
                    }

                    pat
                }
                Pat::Wild(pat) => return Err(ParameterError::WildcardPattern(pat.span())),
                pat => return Err(ParameterError::PatternDestructuring(pat.span())),
            };

            Ok((ident, ty))
        }

        let mut iter = iter.into_iter();
        let (receiver, arg) = {
            let arg = iter.next().ok_or(ParameterError::MissingContext)?;

            match arg {
                FnArg::Receiver(receiver) => {
                    use syn::parse_quote;

                    let and_token = receiver.reference.as_ref().map(|r| r.0);
                    let lifetime = receiver.reference.as_ref().map(|r| r.1.as_ref()).flatten();
                    // Mutability is part of type only when it is a reference.
                    let mutability = and_token.and(receiver.mutability.as_ref());

                    // Make synthetic type for our own use.
                    let ty: Type = parse_quote! {
                        #and_token#lifetime #mutability Self
                    };

                    let receiver = (ReceiverIdent::Receiver(receiver), ty);
                    (Some(receiver), None)
                }
                FnArg::Typed(pat) => {
                    let (ident, ty) = process_pat(pat)?;

                    if ident.ident == "self" {
                        let receiver = (ReceiverIdent::PatIdent(ident), ty);
                        (Some(receiver), None)
                    } else {
                        let arg = (ident, ty);
                        (None, Some(arg))
                    }
                }
            }
        };

        let mut common: Vec<_> = arg
            .map(Ok)
            .into_iter()
            .chain(iter.map(|arg| match arg {
                FnArg::Receiver(_) => Err(ParameterError::UnexpectedSelf(arg.span())),
                FnArg::Typed(pat) => process_pat(pat),
            }))
            .collect::<Result<_, _>>()?;

        let context = common.pop().ok_or(ParameterError::MissingContext)?;

        let r = Parameters {
            receiver,
            common,
            context,
        };

        Ok(r)
    }

    pub fn has_receiver(&self) -> bool {
        self.receiver.is_some()
    }

    pub fn common_len(&self) -> usize {
        self.common.len()
    }

    pub fn receiver_type(&self) -> Option<&Type> {
        self.receiver.as_ref().map(|t| &t.1)
    }

    pub fn context_type(&self) -> &Type {
        &self.context.1
    }

    pub fn common_types(&self) -> impl Iterator<Item = &Type> {
        self.common.iter().map(|t| &t.1)
    }

    pub fn receiver_arg(&self) -> Option<&ReceiverIdent> {
        self.receiver.as_ref().map(|t| &t.0)
    }

    pub fn context_arg(&self) -> &PatIdent {
        &self.context.0
    }

    pub fn common_args(&self) -> impl Iterator<Item = &PatIdent> {
        self.common.iter().map(|t| &t.0)
    }

    pub fn context_ident(&self) -> &Ident {
        &self.context.0.ident
    }

    pub fn common_idents(&self) -> impl Iterator<Item = &Ident> + Clone {
        self.common.iter().map(|t| &t.0.ident)
    }
}
