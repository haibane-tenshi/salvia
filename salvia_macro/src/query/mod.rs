use proc_macro::TokenStream;

use syn::spanned::Spanned;
use syn::{Error, Item, ReturnType, Type};

mod parameter;
mod visitors;

use parameter::{ParameterError, Parameters};

pub fn query_impl(item: Item) -> Result<TokenStream, Error> {
    use quote::{quote, quote_spanned};
    use syn::{ItemFn, Signature};

    let ItemFn {
        attrs,
        vis: host_visibility,
        sig,
        block,
    } = match item {
        Item::Fn(item_fn) => item_fn,
        item => return Err(Error::new(item.span(), "#[query]: function is expected")),
    };

    let Signature {
        constness,
        asyncness,
        unsafety,
        abi,
        fn_token: _,
        ident: host_ident,
        mut generics,
        paren_token,
        inputs,
        variadic,
        output,
    } = sig;

    let output_type: Option<&Type> = match &output {
        ReturnType::Default => None,
        ReturnType::Type(_, ty) => Some(ty),
    };

    // Some sanity checks.
    {
        let mut errs = vec![];

        if let Some(constness) = constness {
            errs.push(Error::new(
                constness.span(),
                "#[query]: cannot be applied to const functions",
            ))
        }

        if let Some(unsafety) = unsafety {
            errs.push(Error::new(
                unsafety.span(),
                "#[query]: cannot be applied to unsafe functions",
            ))
        }

        if let Some(abi) = abi {
            errs.push(Error::new(
                abi.span(),
                "#[query]: cannot be applied to external functions",
            ))
        }

        if let Some(variadic) = variadic {
            errs.push(Error::new(
                variadic.span(),
                "#[query]: cannot be applied to variadic functions",
            ))
        }

        // This provides slightly better error message than default.
        if asyncness.is_none() {
            errs.push(Error::new(
                host_ident.span(),
                "#[query]: can only be applied to async functions",
            ))
        }

        if let Some(output_type) = output_type {
            use syn::visit::Visit;
            use visitors::ImplTraitVisitor;

            let mut visitor: ImplTraitVisitor = Default::default();
            visitor.visit_type(output_type);

            let iter = visitor.0.into_iter()
                .map(|span| Error::new(span,
                                       "#[query]: `impl Trait` syntax is not supported, consider returning concrete type"));

            errs.extend(iter);
        }

        let err = errs.into_iter().reduce(|mut aggr, err| {
            aggr.combine(err);
            aggr
        });

        if let Some(err) = err {
            return Err(err);
        }
    }

    let where_clause = generics.where_clause.take();
    let generic_params = generics;

    let params = Parameters::from_inputs(inputs)
        .map_err(|err| match err {
            ParameterError::MissingContext => Error::new(paren_token.span, "#[query]: function should accept &::salvia::QueryContext as its last parameter"),
            ParameterError::UnexpectedSelf(span) => Error::new(span, "#[query]: unexpected `self` parameter in function"),
            ParameterError::Attribute(span) => Error::new(span, "#[query]: attributes on parameters are not currently supported"),
            ParameterError::PatternDestructuring(span) => Error::new(span, "#[query]: pattern destructuring in parameters is not supported; consider destructuring inside function body instead"),
            ParameterError::WildcardPattern(span) => Error::new(span, "#[query]: wildcard pattern is not supported, consider giving it a name starting with `_`, for ex. `_my_var`; even though your function doesn't need that parameter, query must capture and cache it to correctly function"),
            ParameterError::ImplTrait(span) => Error::new(span, "#[query]: `impl Trait` syntax is not supported, consider turning it into an explicit generic parameter")
        })?;

    // We cannot alias types as some types might be implicit generics.
    // Instead, just generate some shorthands and expand them everywhere.
    #[allow(non_snake_case)]
    let Input = {
        let rendered_receiver_type = params.receiver_type().map(|ty| quote! {#ty,});
        let common_types = params.common_types();

        quote! { (#rendered_receiver_type #(#common_types,)*) }
    };
    #[allow(non_snake_case)]
    let Output = match output_type {
        Some(ty) => quote! { #ty },
        None => quote! { () },
    };

    // Generate hosted function.
    //
    // Signature:
    //      F: FnOnce(Input, QueryContext) -> Ret + Clone + Send + Sync + 'static,
    //      Ret: Future<Output=(Value, QueryContext)> + Send + 'static,
    //
    // We cannot take context by ref because compiler blows up on lifetimes + closures + async,
    // so instead we pass it through.
    let tenant = {
        use syn::Index;

        let tenant_input_pats = params.common_args();
        let tenant_context_pat = params.context_arg();

        let common_types = params.common_types();
        let context_type = params.context_type();

        // Skip `self`: it is both captured by closure and provided as part of input.
        let start = if params.has_receiver() { 1 } else { 0 };
        let end = start + params.common_len();
        let index = (start..end).map(Index::from);

        quote! {
            let __tenant = move |input: #Input, __cx: ::salvia::QueryContext| async move {
                let val = {
                    let (#(#tenant_input_pats,)* #tenant_context_pat,): (#(#common_types,)* #context_type,) = (#(input.#index,)* &__cx,);

                    // Need to silence the warning in case user function never returns.
                    #[allow(unreachable_code)]
                    async move #block.await
                };

                (val, __cx)
            };
        }
    };

    let host = {
        let self_type = params.receiver_type();
        let context_type = params.context_type();
        let common_types = params.common_types();

        let host_self_arg = params.receiver_arg();
        let host_context_arg = params.context_ident();
        let host_common_args = params.common_idents();
        let common_idents = params.common_idents();

        // There doesn't seem to be any good way to interpolate comma only when option is `Some`,
        // so we just prerender relevant bits with comma.
        let rendered_host_self_arg = params.receiver.as_ref().map(|(arg, ty)| {
            use parameter::ReceiverIdent;

            match arg {
                ReceiverIdent::Receiver(receiver) => quote! { #receiver, },
                ReceiverIdent::PatIdent(pat) => quote! { #pat: #ty, },
            }
        });
        let rendered_host_internal_self = host_self_arg.map(|_| quote! {Clone::clone(&self),});

        let self_check = self_type.map(|ty| quote_spanned! {ty.span()=>
            let _ = ::salvia::macro_support::checks::StashableCheck::<#ty>(::std::marker::PhantomData);
            let _ = ::salvia::macro_support::checks::HashCheck::<#ty>(::std::marker::PhantomData);
        });

        let common_checks = params.common_types()
            .map(|ty| quote_spanned! {ty.span()=>
            let _ = ::salvia::macro_support::checks::StashableCheck::<#ty>(::std::marker::PhantomData);
        });

        let manager_checks = params.common_types()
            .map(|ty| quote_spanned! {ty.span()=>
            let _ = ::salvia::macro_support::checks::HashCheck::<#ty>(::std::marker::PhantomData);
        });

        let context_check = quote_spanned! {context_type.span()=>
            let _: ::std::marker::PhantomData::<&::salvia::QueryContext> = ::std::marker::PhantomData::<#context_type>;
        };

        let output_check = quote_spanned! {output_type.span()=>
            let _ = ::salvia::macro_support::checks::StashableCheck::<#Output>(::std::marker::PhantomData);
        };

        quote! {
            #(#attrs )*
            #host_visibility async fn #host_ident #generic_params(#rendered_host_self_arg #(#host_common_args: #common_types,)* #host_context_arg: &::salvia::QueryContext) #output #where_clause {
                #self_check
                #(#common_checks)*
                #(#manager_checks)*
                #context_check
                #output_check

                // Collect input before instantiating any local variables.
                let __input = (#rendered_host_internal_self #(#common_idents,)*);
                let __cx = #host_context_arg;

                #tenant

                ::salvia::node::query::executor(__tenant, __input, __cx).await.unwrap()
            }
        }
    };

    Ok(host.into())
}
