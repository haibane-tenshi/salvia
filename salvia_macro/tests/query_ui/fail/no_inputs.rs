use salvia::query;

#[query]
async fn f() {}

fn main() {}
