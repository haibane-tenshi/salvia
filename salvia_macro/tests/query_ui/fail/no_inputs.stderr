error: #[query]: function should accept &::salvia::QueryContext as its last parameter
 --> $DIR/no_inputs.rs:4:11
  |
4 | async fn f() {}
  |           ^^
